File: 560-subarray-sum-equals-k/subarray-sum-equals-k.java
class Solution {
    public int subarraySum(int[] nums, int k) {
        int count = 0;
        int sum = 0;

        HashMap<Integer, Integer> prefixMap = new HashMap<>();

        prefixMap.put(0, 1);

        for(int num: nums){
            sum += num;
            
            if(prefixMap.containsKey(sum-k)){
                count += prefixMap.get(sum-k);
            }

            prefixMap.put(sum, prefixMap.getOrDefault(sum, 0) + 1);
        }

        return count;
    }
}

File: 560-subarray-sum-equals-k/subarray-sum-equals-k.cpp
class Solution {
public:
    int subarraySum(vector<int>& nums, int k) {
        unordered_map<int, int> mpp;
        int prefixSum = 0, cnt = 0;
        mpp[0] = 1;

        for(int i = 0; i<nums.size(); i++){
            prefixSum += nums[i];
            int remove = prefixSum - k;
            cnt += mpp[remove];
            mpp[prefixSum] += 1;
        }

        return cnt;
    }
};